### ISSUE TITLE
Scrolling behaviour

### ISSUE THREAD

**Issue by pedrorrodriguez** [aurelienpierre/R-Darktable#11]
More of a question than a FR, since perhaps there is an option for this hidden somewhere in the settings. I have always been somewhat frustrated by darktable's scrolling behaviour, and this carries onto R&Darktable. If there is a reason for this I apologize in advance for raising the question. It seems like said behaviour is not logical. Scrolling on the image preview in the darkroom works as expected (zoom in/out). On the side panels however, instead of scrolling up/down, like we have come to expect in every webpage and most other applications, it changes the values of whatever slider you happen to be hovering on. I'd expect ctrl+scrollwheel to alter values, instead of simple scroll. In fact, at the moment we have three different modifiers which affect the speed at which the values change: scrollwheel, ctrl+scrollwheel, alt+scrollwheel, ctrl+shift+scrollwheel. If I want to scroll up or down, I have to hover exactly over the scrollbar (at a very slow speed). If I miss (and I often miss), values are changed to my frustration.

My proposal:
- Simple scrolling on all panels should scroll the panel up or down
- Ctrl+scroll (or whichever other modifier you prefer) actually changes slider values.
- Shift+scroll, alt+scroll and/or ctrl+alt+scroll for precision value changes if we still need three scrolling speeds.
- The current shift+scroll behavior I find useless. For large value changes we can simple click or click and drag on the slider, and for incredibly precise value changes we still have right click with the tail thingy.

Should I try and take this upstream, @aurelienpierre or leave this here?

**Comment by StudioPetrikas** [comment]
I've messed up a lot of photos by scrolling, before finding this setting:
![sc-2022-07-07 00_58_31](https://user-images.githubusercontent.com/70055734/177656997-6f558236-052b-4ea8-a9a3-e2d68dc842f4.jpg)

Should be on by default, IMO.

**Comment by pedrorrodriguez** [comment]
> The general consideration here is that Ctrl+Scroll is more cumbersome than just Scroll, so it should be affected to the operation you realize the less often. 
> Being able to scroll directly on top of sliders is a time saver, and also makes it possible to edit without a keyboard. I think relying too heavily on key modifiers is just another flavour of not good, though I hear that the current behaviour is inconsistent with the rest of the world.


I'd be fine with ctrl+scroll for vertical scrolling even though it is inconsistent with the rest of the world if it does save time. To me, the annoyance of accidentally messing up edits trumps the slight inconvenience of the modifier key. My problem is not having any sort of modifier option to vertically scroll. 

> Now, if you consistently miss the scrollbar, that may indicate another underlying problem : it's too narrow.

True, but having to mouse over and scroll to me seems more cumbersome than a modifier key, isn't it?

**Comment by aurelienpierre** [comment]
> True, but having to mouse over and scroll to me seems more cumbersome than a modifier key, isn't it?

Depends on the distance you have to travel. 

I think here we have to zoom out on the present problem and consider the more general problem: how to assign and distribute scrolling events to GUI elements in sidebar to promote efficient interaction and minimal mistakes ? Let's not rush into shortcuts and key mapping too fast.

**Comment by StudioPetrikas** [comment]
> Now, if you consistently miss the scrollbar, that may indicate another underlying problem : it's too narrow.

I think it's too small, but, if working full screen, one can just bounce the cursor to the right of the screen and scroll. So thickness isn't the main issue, IMO.

I'd love if there was a possibility to middle click and drag to scroll. More often, one only needs to scroll till the start/end of module or module list. Rarely do you go full wheel strokes to scroll somewhere. So middle-click and drag would be viable, me thinks.

**Comment by pedrorrodriguez** [comment]
> > Now, if you consistently miss the scrollbar, that may indicate another underlying problem : it's too narrow.
> 
> I think it's too small, but, if working full screen, one can just bounce the cursor to the right of the screen and scroll. So thickness isn't the main issue, IMO.

If you are working on a laptop it is. Mine has a very nice trackpad, and I still miss quite often.

> I'd love if there was a possibility to middle click and drag to scroll. More often, one only needs to scroll till the start/end of module or module list. Rarely do you go full wheel strokes to scroll somewhere. So middle-click and drag would be viable, me thinks.

It's not a huge issue, and the checkbox you showed me actually makes things work for me. But when I wasn't aware of it, it was tiresome to go between exposure, filmic, tone equalizer, color balance rgb... Also on the left panel, if you leave open the snapshot module, history, color picker and export you face the same issue.

Middle click and drag can work, although when working from a laptop that usually means a modifier key+scroll or similar.

**Comment by aurelienpierre** [comment]
This issue will be a `won't fix` until you bear with me and follow the proper steps of design. We already have too many proposed solutions (middle click + drag, make Ctrl+Scroll a default for sliders, remove scroll on sliders entirely, etc.) and yet no complete understanding of the problem. 

I need to know:

1. what your usual setup is : 
    * laptop or desktop,
    * mouse (2 or 3 buttons ?) or Wacom or keyboard or a combination of those,
2.  what the most frequent operation you are doing is:
    * scrolling vertically in sidebar content ?
    * pushing sliders ?
    * roughly 50/50 ?
3. if you are aware that Shift+Click on module headers collapses all modules but the one you clicked AND aligns it in the sidebar (possible easy fix to the problem here: no scrolling needed),
4. any other **problem** you have.

Please keep away from solutions, they are not helping.

**Comment by StudioPetrikas** [comment]
I'll join in just because I would like to see this improved. 
I think Resolve has done an excellent job in this regard - I've never had an issue. So if anyone is looking for a good UX example for sidebar tools, I think it's a good start. Notably, sliders change values by scrolling ONLY after clicked. (Usually, the first adjustment is not always exact, you jab in the ballpark and then fine-adjust it). Middle-click and dragging is also present. As it is in Blender.


    **1. What your usual setup is :**
        Laptop with Touchpad when on the move; External Keyboard + Mouse when stationary.
        Normal mouse from this century, with a clicking scroll wheel.
    **2. What the most frequent operation you are doing is:**
        50/50. There's a lot of scrolling and adjusting at the beginning of the workflow. Less scrolling towards the end of development.
    **3. If you are aware that Shift+Click on module headers collapses all modules but the one you clicked AND aligns it in the sidebar (possible easy fix to the problem here: no scrolling needed),**
         Scrolling is still needed, as some of the modules take more than a full height of the screen. I usually work on 2560x1600 screen, and if you expand "color balance rgb" with the mask section expanded, it won't fit; meaning that any module bellow is no longer shift-clickable. You're required to either scroll or hot-key the searchbar and enter whatever. OR collapse the module, but that defeats the purpose of completely removing the ability to have more than 1 module open at a time.
    **4. How ofter misclicks and mistakes happen**
    Very often. I'm not very "mechanically precise". So bouncing the mouse to the side of the screen and scrolling, I sometimes accidentally move the mouse off the scrollbar and change a value unknowingly. The setting that I showed in a screenshot above helped with that, but, as you said, it's a pain to have the other hand occupied just for clicking control.
    **5. any other problem you have.**
    Scrolling is also very jarring, compared to clicking and dragging. It jumps roughly 3 sliders up/down with the settings I have on my mouse. I often lose track of where I and have to put effort into reorienting. By clicking-and-dragging you're not "disconnected" from where you were before you started moving.
    
Hope this is helpful.

**Comment by aurelienpierre** [comment]
> Notably, sliders change values by scrolling ONLY after clicked.

Yep, I had this one in a corner of my mind since the beginning.

Thanks for playing with me. So for now, I propose to have sliders capture scroll only after having got the focus (that is, after a "void" click on the slider name or right after a previous setting interaction with sliders through key shortcuts or mouse events). The benefit of this is we don't need shady shortcuts and we don't even need keyboard at all. Since scrolling is done with a mouse, then everything is done with the mouse (giving focus and interacting). 

Now, since you both mention:

> Scrolling is also very jarring, compared to clicking and dragging. It jumps roughly 3 sliders up/down with the settings I have on my mouse. I often lose track of where I and have to put effort into reorienting. By clicking-and-dragging you're not "disconnected" from where you were before you started moving.

> The scrolling speed when hovering over the scrollbar is slow and I don't know how to change it. With the checkbox that was suggested above checked, scrolling over the side bars (where the modules are placed) happens at a normal speed.

We have another problem of vertical scrolling. I have to check, but the number of lines scrolled at each wheel increment is supposed to be set by Gtk and might depend on your desktop environment. There is probably a way to hard-set it at the app level.

**Comment by pedrorrodriguez** [comment]
Sounds great, thanks! It seems like a good solution. Looking forward to testing it.

Btw, what you did, telling us exactly what you needed was very helpful. It's not what github recommends when submitting a feature request, though. You might want to put a list like that together that you (and others) can point to, so that feedback can lead to a more useful discussion.

**Comment by aurelienpierre** [comment]
> Btw, what you did, telling us exactly what you needed was very helpful. It's not what github recommends when submitting a feature request, though. You might want to put a list like that together that you (and others) can point to, so that feedback can lead to a more useful discussion.

What I need depends on the nature of the issue though.

**Comment by pedrorrodriguez** [comment]
> > Btw, what you did, telling us exactly what you needed was very helpful. It's not what github recommends when submitting a feature request, though. You might want to put a list like that together that you (and others) can point to, so that feedback can lead to a more useful discussion.
> 
> What I need depends on the nature of the issue though.

I understand that. However, since you have expressed frustration at feedback from users in the form of ill thought solutions, mockups and the like before trying to properly identify the core issue or design goal, there has to be a way for us to give feedback without making you want to pull your hair out. This is what github tells us to do:

**Is your feature request related to a problem? Please describe.**



**Describe the solution you'd like**



**Alternatives**



**Additional context**


It's more or less what I did in the original post, but that wasn't the kind of feedback you needed/wanted. It doesn't have to be a ten page manifesto, but some concise guidelines for people to give feedback would be very helpful.

**Comment by aurelienpierre** [comment]
Ok, I updated it. We will see.

If we are to institute a proper culture of problem solving, I wrote a couple of articles on the topic:
* https://eng.aurelienpierre.com/2020/04/the-designer-and-the-drilling-machine/
* https://eng.aurelienpierre.com/2022/06/design-by-committee-will-not-save-floss/

These describe the challenges I'm facing in this job, and I'm a big believer in understanding your coworker's challenges in order to better interact with them.

**Comment by pedrorrodriguez** [comment]
Thank you! I'll give those a read. Hopefully this gentle reminder every time we open a new issue will help.

**Comment by aurelienpierre** [comment]
Planned for tomorrow.

**Comment by aurelienpierre** [comment]
List of checks:

- [ ] the sidebar (Gtk scrollable window) captures the scroll if no slider got the focus
- [ ] clicking on the slider name gives focus without changing the value
- [ ] the slider still has enough "grace" space around to allow sloppy clicking still setting the value
- [ ] clicking in a void space removes the focus from the last used slider

**Comment by aurelienpierre** [comment]
Please test #23 or `git checkout candidate` and check the list of tasks above (the last item is not implemented, it's more than an afternoon job – still trying to assess if it's worth it).

**Comment by StudioPetrikas** [comment]
- [ ] 1. Sort of. If no slider has focus, the scroll wheel scrolls through the tabs in the module, once on the last tab, then the whole module pane. Eg: if I scroll on an empty space in the "Filmic rgb" module, it goes from Scene, to Reconstruct, to Look, to Display, to Options and then, once at the options tab, starts scrolling the module down. This is also inconsistent. Sometimes a slider grabs attention without being clicked. Sometimes it goes to a different tab. Sometimes it scrolls up and down. There might be an issue created by hovering on the tab. Non-tabbed modules behave correctly.
- [ ] 2. Sort of. Clicking on the slider area does give it focus. But some sliders still capture focus unclicked (on hover)
- [x ] 3. 100%. Feels great!
- [x] 4. Works well from my short testing
- [ ] 5.

Scrolling smoothness was improved drastically. I'm no longer disoriented and can navigate more easily. Solid step forward, thank you!

**Comment by aurelienpierre** [comment]
> If no slider has focus, the scroll wheel scrolls through the tabs in the module, once on the last tab, then the whole module pane

That's how I learn that someone found clever to add mouse wheel support for tabs. Another idea that tried to be cool and ended up solving nothing but creating problems. That shit goes now.

**Comment by aurelienpierre** [comment]
Dunno, but it's gone with latest commit.

**Comment by StudioPetrikas** [comment]
Okay, updates:

- [x] 1. Perfect.
- [ ] 2. It appears that every top slider of the 'filmic rgb' module captures scrolling unclicked. The lower sliders don't. Same for "color balance rgb". Basically, every tabbed module. It seems that the click gets capture from clicking the tab.
- [x] 3. 100%. Feels great!
- [x] 4. Works well from my short testing
- [ ] 5.

**Comment by pedrorrodriguez** [comment]
Everything works perfectly. Regarding scrolling speed, more precise feedback:

- Fine on the sidebar (on the modules), but slightly on the fast side (perhaps like @StudioPetrikas, although since I'm using a touchpad, it's fine?)
- Slow on the scrollbar.

**Comment by aurelienpierre** [comment]
> It appears that every top slider of the 'filmic rgb' module captures scrolling unclicked. The lower sliders don't. Same for "color balance rgb". Basically, every tabbed module. It seems that the click gets capture from clicking the tab.

Ok, apparently Gtk automatically gives the focus to the first child widget in the tab when you enable it. I just have to override it manually.

**Comment by aurelienpierre** [comment]
> Fine on the sidebar (on the modules), but slightly on the fast side (perhaps like @StudioPetrikas, although since I'm using a touchpad, it's fine?)
    Slow on the scrollbar.

Wow, ok, I never realized the scrolling speed was not the same on the scrollbar and on the content. This is for another day though (#16)

**Comment by StudioPetrikas** [comment]
@pedrorrodriguez I would say that the current (after the latest update) scroll speed on the scrollbar is ideal and too fast on the sidebar, would you agree?

I suppose the question is, how often do we need to scroll "far" vs "near".

I find myself needing to do small scrolling adjustments, just to reach the module title bar in order to expand/collapse/focus it. Or scroll down slightly to reach the mask section. Rarely do I need to jump a mile down (and if I do, I then grab the scroll-bar)

Thoughts please.

**Comment by pedrorrodriguez** [comment]
@StudioPetrikas, as you saw, the scrolling speed on the scrollbar is extremely slow, although I agree with you, now that scrolling on the sidebar works well, it might not be needed anymore. Regarding the scrolling speed on the sidebar, it is a bit on the fast side for me, but usable. 

I also agree with you, that generally I just need small scrolling adjustments.

**Comment by aurelienpierre** [comment]
Switching tabs should not give focus to the first slider anymore with the latest commit.

**Comment by StudioPetrikas** [comment]
I'm no longer able to test the "fix-scrolling" branch. 
I can no longer load any file into darkroom. I can select a file and switch to darkroom, where it hangs on "loading [filename"... and a gray background and becomes unresponsive.

Tried with different filetypes. Tried building from scratch. 

Windows 11
3.9.0+1714~g0c045929a

Last known working commit was 1c2f4907b0bcd2563f366fb9e53669debc180c3c "Enlarge Scroll bars in GUI"

**Comment by aurelienpierre** [comment]
@StudioPetrikas I reverted the responsible commit just now, should work again.

**Comment by StudioPetrikas** [comment]
Thank you for that! 

- [x] 1. Perfect.
- [x] 2. Works well.
- [x] 3. 100%. Feels great!
- [x] 4. Works well from my short testing
- [ ] 5.

So the only weirdness that's left is that the scrollbar scrolling behaves differently to sidebar scrolling. Scrollbar scrolling feels super lovely on a mousewheel, but is dead slow on a touchpad. Sidebar scrolling is still too fast on a mousewheel, but feels just right on the toucpad. Can't offer a possible solution, because I don't have one.

Clicking the wheel and dragging (grab-and-drag) would still be the best for mouse users, but that's a different topic and a different feature-request.

Thanks for this MASSIVE improvement!

**Comment by aurelienpierre** [comment]
>  2. Switching module tabs grabs the attention of the first slider

I don't understand.

> Yes, but. If I click on an empty area and scroll on the sidebar, the module list scrolls up or down, except if my cursor goes over the last selected slider. In that case, it will still grab the focus and change the values.

Will investigate tomorrow.

**Comment by StudioPetrikas** [comment]
![Animation2](https://user-images.githubusercontent.com/70055734/180330338-860372d0-7104-44bb-b4ba-ddbdae9b1fcc.gif)

Version: 3.9.0+1715~g97a29993e

**Comment by aurelienpierre** [comment]
Shit, I thought I fixed that…

**Comment by aurelienpierre** [comment]
Ok, I have entirely changed the logic and simplified it with an internal handling instead of relying on Gtk events.

All issues should be fixed except 5.

**Comment by StudioPetrikas** [comment]
Awesome! Just tested, everything works well; tabbed module issue fixed.
Let me know when further testing is needed.

**Comment by pedrorrodriguez** [comment]
Sorry that I haven't been able to get back to you until now. Everything works well on my end. Thank you for the fix! Let us know if more testing is needed.
